imap_processing/mag/l1a/mag_l1a_data.py

```diff
@@ -429,6 +477,467 @@ def to_signed16(n: int) -> int:
+            updated_vectors[i + 1][3] = range_int
+        return updated_vectors
+
+    @staticmethod
+    def accumulate_vectors(
+        first_vector: np.ndarray,
+        vector_differences: list[int],
+        vector_count: int,
+    ) -> np.ndarray:
+        """
+        Given a list of differences and the first vector, return calculated vectors.
+
+        This is calculated as follows:
+            vector[i][0] = vector[i-1][0] + vector_differences[i][0]
+            vector[i][1] = vector[i-1][1] + vector_differences[i][1]
+            vector[i][2] = vector[i-1][2] + vector_differences[i][2]
+            vector[i][3] = first_vector[3]
+
+        The third element of the array is the range value, which we assume is the same
+        as the first vector.
+
+        Parameters
+        ----------
+        first_vector : numpy.ndarray
+            A numpy array of 3 signed integers and a range value, representing the
+            start vector.
+        vector_differences : numpy.ndarray
+            A numpy array of shape (expected_vector_count, 4) of signed integers,
+            representing the differences between vectors.
+        vector_count : int
+            The expected number of vectors in the output.
+
+        Returns
+        -------
+        numpy.ndarray
+            A numpy array of shape (expected_vector_count, 4) of signed integers,
+            representing the calculated vectors.
+        """
+        vectors: np.ndarray = np.empty((vector_count, 4), dtype=np.int32)
+        vectors[0] = first_vector
+
+        index = 0
+        vector_index = 1
+        for diff in vector_differences:
+            vectors[vector_index][index] = vectors[vector_index - 1][index] + diff
+            index += 1
+            if index == 3:
+                # Update range section to match that of the first vector
+                vectors[vector_index][3] = vectors[0][3]
+                index = 0
+                vector_index += 1
+
+        return vectors
+
+    @staticmethod
+    def unpack_one_vector(
+        vector_data: np.ndarray, width: int, has_range: int
+    ) -> np.ndarray:
+        """
+        Unpack a single vector from the vector data.
+
+        Input should be a numpy array of bits, eg [0, 0, 0, 1], of the length width*3,
+        or width*3 + 2 if has_range is True.
+
+        Parameters
+        ----------
+        vector_data : numpy.ndarray
+            Vector data for the vector to unpack. This is uncompressed data as a numpy
+            array of bits (the output of np.unpackbits).
+        width : int
+            The width of each vector component in bits. This needs to be a multiple of
+            8 (including only whole bytes).
+        has_range : int
+            1 if the vector data includes range data, 0 if not. The first vector always
+            has range data, it is only if the compression fails and we revert to
+            uncompressed data partway through that we skip the range.
```

Comment:
So would it either be one of two scenarios:
1. compressed - range data
2. compressed + uncompressed - range data
3. uncompressed - no range data

What do you mean by "partway through"? This is more for my own understanding than anything.

Reply:
You're correct. I'll reword this sentence because I understand your confusion. The `has_range` section is totally separate from the compressed/uncompressed bits. What this sentence is trying to say, is that the uncompressed data does not include a range value. But that should probably be said somewhere else. 

It is also possible to have compressed + uncompressed with no range data. If the range stays the same for the entire packet, then we do not have a range data section. Otherwise, we retrieve all the range data for all the vectors from the range data, regardless if the vectors are compressed or uncompressed. 

I'll rewrite this sentence because it is confusingly tying things together when they should be separate, thanks for calling that out!